File: Array/二叉搜索树的第k个结点.cpp
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};
*/
class Solution {
public:
    TreeNode* KthNode(TreeNode* pRoot, unsigned int k)
    {
        if(pRoot == nullptr || k==0){
            return nullptr;
        }
        return KthNodeCore(pRoot, k);

    }
    TreeNode* KthNodeCore(TreeNode* pRoot, unsigned int& k){
        TreeNode* target = nullptr;
        if(pRoot->left != nullptr){
            target = KthNodeCore(pRoot->left, k);
        }
        if(target == nullptr){
            if(k == 1){
                return pRoot;
            }
            k--;
        }
        if(target == nullptr && pRoot->right != nullptr){
            target = KthNodeCore(pRoot->right, k);
        }
        return target;
    }
    
};

File: Array/二叉搜索树的第k个结点.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    # 返回对应节点TreeNode
    def __init__(self):
        self.count = 0
    def KthNode(self, pRoot, k):
        # write code here
        if pRoot:
            target = self.KthNode(pRoot.left, k)
            if target:
                return target
            self.count += 1
            if self.count == k:
                return pRoot
            target = self.KthNode(pRoot.right, k)
            if target:
                return target
        else:
            return None
        